superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx

```diff
@@ -36,16 +41,17 @@ export const dndGroupByControl: SharedControlConfig<'DndColumnSelect'> = {
   ),
   mapStateToProps(state, { includeTime }) {
     const newState: ExtraControlProps = {};
-    if (state.datasource) {
-      const options = state.datasource.columns.filter(c => c.groupby);
+    const { datasource } = state;
+    if (datasource?.columns[0]?.hasOwnProperty('groupby')) {
+      const options = (datasource as Dataset).columns.filter(c => c.groupby);
       if (includeTime) {
         options.unshift(TIME_COLUMN_OPTION);
       }
       newState.options = Object.fromEntries(
         options.map(option => [option.column_name, option]),
       );
-      newState.savedMetrics = state.datasource.metrics || [];
-    }
+      newState.savedMetrics = (datasource as Dataset).metrics || [];
+    } else newState.options = datasource?.columns;
```

Comment:
@lyndsiWilliams I think we're going to need lines 47-52 applied to query columns as well. Just the filter we can skip.

Reply:
Done in [`this commit`](https://github.com/apache/superset/pull/19855/commits/ccd3e3e800b99db12bb2117793a77a243c60240d) 😁